libraries/AP_L1_Control/AP_L1_Control.cpp

```diff
@@ -202,11 +208,40 @@ void AP_L1_Control::update_waypoint(const struct Location &prev_WP, const struct
 		ltrackVel = _groundspeed_vector * AB; // Velocity along track
 		float Nu2 = atan2f(xtrackVel,ltrackVel);
 		//Calculate Nu1 angle (Angle to L1 reference point)
-		float xtrackErr = A_air % AB;
-		float sine_Nu1 = xtrackErr/max(_L1_dist, 0.1f);
+		float sine_Nu1 = _crosstrack_error/max(_L1_dist, 0.1f);
 		//Limit sine of Nu1 to provide a controlled track capture angle of 45 deg
 		sine_Nu1 = constrain_float(sine_Nu1, -0.7071f, 0.7071f);
 		float Nu1 = asinf(sine_Nu1);
+
+        // compute integral error component to converge to a crosstrack of zero
+		// only when traveling straight and level
+		if (fabsf(Nu1) < radians(5)) {
+
+            // to converge to zero we must push Nu1 harder but only compute
+            const float i_gain = 5.0e-3f; // a very very slow gain, takes several minutes to settle
+            const float dt = 0.01f; // 10Hz
+            _L1_xtrack_integrator += Nu1 * i_gain * dt;
+
+            // an AHRS_TRIM_X=0.1 will drift to about 0.08 so 0.1 is a good worst-case to clip at
+            _L1_xtrack_integrator = constrain_float(_L1_xtrack_integrator, -0.1f, 0.1f);
+
+            // store _L1_xtrack_integrator to non-volatile memory if
+            // it has changed more than 10% but no sooner than 5 minutes since last write
+            uint32_t now_ms = hal.scheduler->millis();
+            if (now_ms - _L1_xtrack_integrator_last_write_ms > 5*60*1000) {
```

Comment:
I'm quite sceptical about the need to save this value. It assumes that whatever is causing the xtrack error is a persistent property of the airframe. That is true if its a autopilot install error, but isn't true for other causes. Some of the other causes may result in the xtrack being in earth frame not body frame (eg. wind), so saving doesn't make sense, and in fact resetting on turns may make sense.